tests/unit/s2n_key_update_threads_test.c

```diff
@@ -0,0 +1,247 @@
+            bytes_to_send -= r;
+        } else if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            fprintf(stderr, "Send error: %s. %s\n", s2n_strerror(s2n_errno, NULL), s2n_strerror_debug(s2n_errno, NULL));
+            return NULL;
+        }
+    }
+    return conn;
+}
+
+static void *s2n_recv_random_data(void *arg)
+{
+    struct s2n_connection *conn = (struct s2n_connection *) arg;
+
+    uint8_t buffer[S2N_TEST_BUFFER_SIZE] = { 0 };
+
+    size_t bytes_to_read = S2N_TEST_BYTES_TO_SEND;
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    while (bytes_to_read) {
+        int r = s2n_recv(conn, buffer, MIN(sizeof(buffer), bytes_to_read), &blocked);
+        if (r >= 0) {
+            bytes_to_read -= r;
+        } else if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            fprintf(stderr, "Recv error: %s. %s\n", s2n_strerror(s2n_errno, NULL), s2n_strerror_debug(s2n_errno, NULL));
+            return NULL;
+        }
+    }
+    return conn;
+}
+
+static S2N_RESULT s2n_send_and_recv_random_data(struct s2n_connection *conn)
+{
+    /*
+     * This test is intended to find concurrency issues when sending and receiving
+     * KeyUpdates, so we need to run the reader and writer in separate threads.
+     */
+
+    pthread_t reader = 0;
+    RESULT_ENSURE_EQ(pthread_create(&reader, NULL, s2n_recv_random_data, (void *) conn), 0);
+
+    pthread_t writer = 0;
+    RESULT_ENSURE_EQ(pthread_create(&writer, NULL, s2n_send_random_data, (void *) conn), 0);
+
+    void *reader_return = NULL;
+    RESULT_ENSURE_EQ(pthread_join(reader, &reader_return), 0);
+    RESULT_ENSURE_REF(reader_return);
+
+    void *writer_return = NULL;
+    RESULT_ENSURE_EQ(pthread_join(writer, &writer_return), 0);
+    RESULT_ENSURE_REF(writer_return);
+
+    RESULT_ENSURE_GT(conn->wire_bytes_out, S2N_TEST_BYTES_TO_SEND);
+    RESULT_ENSURE_GT(conn->wire_bytes_in, S2N_TEST_BYTES_TO_SEND);
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_sanity_check_key_updates_sent(struct s2n_connection *conn)
+{
+    struct s2n_blob seq_num_blob = { 0 };
+    if (conn->mode == S2N_CLIENT) {
+        RESULT_GUARD_POSIX(s2n_blob_init(&seq_num_blob, conn->secure->client_sequence_number,
+                sizeof(conn->secure->client_sequence_number)));
+    } else {
+        RESULT_GUARD_POSIX(s2n_blob_init(&seq_num_blob, conn->secure->server_sequence_number,
+                sizeof(conn->secure->server_sequence_number)));
+    }
+
+    uint64_t seq_num = 0;
+    RESULT_GUARD_POSIX(s2n_sequence_number_to_uint64(&seq_num_blob, &seq_num));
+    RESULT_ENSURE_LTE(seq_num, conn->secure->cipher_suite->record_alg->encryption_limit);
+
+    /* Verify that if no KeyUpdate was sent and seq_num therefore represents the
+     * total number of records sent, we cannot account for all of wire-bytes_out.
+     * This proves that at least one KeyUpdate was sent.
+     */
+    RESULT_ENSURE_GT(conn->wire_bytes_out, S2N_DEFAULT_RECORD_LENGTH * seq_num);
```

Comment:
I'm still sort of struggling to understand this key update sanity check function and why it proves that a key update was done.

Reply:
I updated the comment. But I'm wondering if we should just add a key_update counter to the connection. Really just for testing; in production we couldn't assume it wouldn't overflow / wrap.